labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java

```diff
@@ -136,6 +139,59 @@ public void createLabel(
     }
   }
 
+  @Override
+  public void bulkCreateLabels(
+      BulkCreateLabelsRequest bulkCreateLabelsRequest,
+      StreamObserver<BulkCreateLabelsResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      Lock labelsLock = this.stripedLabelsLock.get(requestContext.getTenantId());
+      if (labelsLock.tryLock(WAIT_TIME.getSeconds(), TimeUnit.SECONDS)) {
+        try {
+          List<Label> labels =
+              bulkCreateLabelsRequest.getRequestsList().stream()
+                  .map(this::buildLabelFromRequest)
+                  .filter(Optional::isPresent)
+                  .map(Optional::get)
+                  .collect(Collectors.toList());
+          if (!labels.isEmpty()) {
+            List<Label> createdLabels =
+                labelStore.upsertObjects(requestContext, labels).stream()
+                    .map(org.hypertrace.config.objectstore.ConfigObject::getData)
+                    .collect(Collectors.toList());
+            responseObserver.onNext(
+                BulkCreateLabelsResponse.newBuilder().addAllLabels(createdLabels).build());
+          } else {
+            responseObserver.onNext(BulkCreateLabelsResponse.newBuilder().build());
+          }
+          responseObserver.onCompleted();
+        } finally {
+          labelsLock.unlock();
+        }
+      } else {
+        responseObserver.onError(new StatusRuntimeException(Status.ABORTED));
+      }
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  private Optional<Label> buildLabelFromRequest(BulkCreateLabelsRequest.LabelRequest request) {
+    LabelData labelData = request.getData();
+    if (systemLabelsKeyLabelMap.containsKey(labelData.getKey())
```

Comment:
I'm tempted to error if they try to create something that already exists - at least on system labels, where there's no worry on races. On user labels, we may still end up in an inconsistent state if we don't error this, since the caller may thing it was created with the parameter they specified and it was actually created with some other parameters. This is particularly problematic if we allow the caller to pre-define IDs so they don't have to block on the response. Also, would this cause the existing label to be omitted from the response? that's problematic too - the caller wouldn't be able to get IDs back

Reply:
I had ignored duplicate labels for the case when two enricher instance is trying to create duplicate labels as part of bulk create. Handling error will be difficult in that case. We will have bail out without creating labels.

Reply:
Yeah, handling that case is important (although what's the partitioning like? If it's by tenant then that would greatly mitigate this concern, right).

So I think we need to think about the pre-defining IDs case - even if we don't do it yet - because that may influence how this gets built. If two clients are asking to create the same label, what does that ask look like and what does the response look like for the one that arrives first and the one that arrives second?

Reply:
Discussed this offline and conclusion is that for existing labels, we will still return them as part of response.